proof/crefine/ARM_HYP/Wellformed_C.thy

```diff
@@ -271,8 +271,70 @@ definition
                    | Some cap \<Rightarrow> Some \<lparr> cap_CL = cap,
                                        cteMDBNode_CL = mdb_node_lift (cteMDBNode_C c) \<rparr>"
 
-lemma to_bool_false [simp]: "\<not> to_bool false"
-  by (simp add: to_bool_def false_def)
+lemma true_and_1[simp]:
+  "true && 1 = true"
+  "signed true && 1 = signed true"
+  "unsigned true && 1 = unsigned true"
+  by (simp add: true_def)+
+
+lemma to_bool_true[simp]:
+  "to_bool true"
+  "to_bool (signed true)"
+  "to_bool (unsigned true)"
+  by (simp add: to_bool_def true_def)+
+
+lemma true_equals_simps[simp]:
+  "true = 1 \<longleftrightarrow> True"
+  "signed true = 1 \<longleftrightarrow> True"
+  "unsigned true = 1 \<longleftrightarrow> True"
+  "true = 0 \<longleftrightarrow> False"
+  "signed true = 0 \<longleftrightarrow> False"
+  "unsigned true = 0 \<longleftrightarrow> False"
+  by (simp add: true_def)+
+
+lemma true_eq_from_bool[simp]:
+  "(true = from_bool P) = P"
+  "(signed true = from_bool P) = P"
+  "(unsigned true = from_bool P) = P"
+  by (simp add: from_bool_def split: bool.splits)+
+
+lemma false_and_1[simp]:
+  "false && 1 = false"
+  "signed false && 1 = signed false"
+  "unsigned false && 1 = unsigned false"
+  by (simp add: false_def)+
+
+lemma to_bool_false[simp]:
+  "\<not> to_bool false"
+  "\<not> to_bool (signed false)"
+  "\<not> to_bool (unsigned false)"
+  by (simp add: to_bool_def false_def)+
+
+lemma false_equals_simps[simp]:
+  "false = 0 \<longleftrightarrow> True"
+  "signed false = 0 \<longleftrightarrow> True"
+  "unsigned false = 0 \<longleftrightarrow> True"
+  "false = 1 \<longleftrightarrow> False"
+  "signed false = 1 \<longleftrightarrow> False"
+  "unsigned false = 1 \<longleftrightarrow> False"
+  by (simp add: false_def)+
+
+lemma false_eq_from_bool[simp]:
+  "(false = from_bool P) = (\<not> P)"
+  "(signed false = from_bool P) = (\<not> P)"
+  "(unsigned false = from_bool P) = (\<not> P)"
+  by (simp add: from_bool_def split: bool.splits)+
+
+lemma from_bool_vals[simp]:
+  "from_bool True = signed true"
+  "from_bool False = signed false"
+  by (simp add: from_bool_def)+
+
+lemma true_neq_false[simp]:
+  "true \<noteq> false"
+  "signed true \<noteq> signed false"
+  "unsigned true \<noteq> unsigned false"
+  by (simp add: true_def false_def)+
```

Comment:
These get repeated across all `Wellformed_C.thy` files (except X64, does that not have a `Wellformed_C`?
I've got a feeling that these should not be duplicated like that. Can they go into `proof/crefine/lib` somehow? There's stuff in there that imports SR_Lemmas, so post-kernel-parse things should fit.

Reply:
Oh, and they are also repeated in X64's `Wellformed_C`. That file might have been hiding from you when scrolling through the list, it's right at the bottom.

Reply:
Yup, missed it!
New file would work for me. First idea that comes to mind is `Boolean_C` which could accommodate the `to_bool_bf`, but it'll be easy to rename if people get a better idea

Reply:
> Yup, missed it! New file would work for me. First idea that comes to mind is `Boolean_C` which could accommodate the `to_bool_bf`, but it'll be easy to rename if people get a better idea

This is done now, thanks for the suggestion! This is a lot nicer with the reduced duplication.

Reply:
Looks a lot nicer, and another small step forward to a more generic CRefine!
Maybe Gerwin has some crazy idea to make Boolean_C *really* generic by using locales, but I'm happy with it as is.

Reply:
Happy with it as is.